### PR TITLE
OF-862 Add support for XEP-313: Message Archive Management

### PR DESCRIPTION
**Pull request by JonnyHeavey**
Implementation of MAM to add user message archives.

### CODE REVIEW

**Comment by deleolajide** [comment]
Excellent work :-) 

Could you kindly remove these files for the openfire meetings (ofmeet) plugin. They will overwrite the master versions.

src/plugins/ofmeet/src/WEB-INF/classes/org/jitsi/videobridge/openfire/Config.class
src/plugins/ofmeet/src/WEB-INF/classes/org/jitsi/videobridge/openfire/HttpProxy.class
src/plugins/ofmeet/src/WEB-INF/classes/org/jivesoftware/openfire/plugin/spark/Bookmark$Type.class
src/plugins/ofmeet/src/WEB-INF/classes/org/jivesoftware/openfire/plugin/spark/Bookmark.class
src/plugins/ofmeet/src/WEB-INF/classes/org/jivesoftware/openfire/plugin/spark/BookmarkManager.class

**Comment by deleolajide** [comment]
Why is this PR referencing files in the ofmeet plugin?

**Comment by JonnyHeavey** [comment]
They were accidentally included, I have now reverted (removed) the changes from the PR

**Comment by deleolajide** [comment]
Thanks. Exceļlent Contribution to Openfire. 
+1 from me

**Comment by Redor** [comment]
+1, nice work

File: src/plugins/monitoring/src/java/org/jivesoftware/openfire/archive/ConversationManager.java [hunk 1]
```diff
@@ -85,8 +85,8 @@
 
 	private static final String UPDATE_CONVERSATION = "UPDATE ofConversation SET lastActivity=?, messageCount=? WHERE conversationID=?";
 	private static final String UPDATE_PARTICIPANT = "UPDATE ofConParticipant SET leftDate=? WHERE conversationID=? AND bareJID=? AND jidResource=? AND joinedDate=?";
-	private static final String INSERT_MESSAGE = "INSERT INTO ofMessageArchive(conversationID, fromJID, fromJIDResource, toJID, toJIDResource, sentDate, body) "
-			+ "VALUES (?,?,?,?,?,?,?)";
+	private static final String INSERT_MESSAGE = "INSERT INTO ofMessageArchive(messageID, conversationID, fromJID, fromJIDResource, toJID, toJIDResource, sentDate, body, stanza) "
+			+ "VALUES ((SELECT COUNT(*) FROM ofMessageArchive),?,?,?,?,?,?,?,?)";
```
Comment (maheshpec) [inline src/plugins/monitoring/src/java/org/jivesoftware/openfire/archive/ConversationManager.java]: Hi Jonny,
This script is failing in mysql. I've opened up a discussion for this in the forum - https://community.igniterealtime.org/thread/55794.
Thanks for commiting this change :+1: 
Mahesh

**Comment by mrunalpittalia** [comment]
Hi,

I am using Openfire 3.10.2 latest release (using MySQL database type) with installed "Monitoring Service v 1.4.6 plugin" having support for MAM.

However, openfire sends "feature-not-implemented" in response of getting mam pref call.

* Here below is stanza used to get he mam pref. from openfire server:
`<body rid="550948740" xmlns="http://jabber.org/protocol/httpbind" sid="9c644adb">  
     <iq id="1:sendIQ" type="get" xmlns="jabber:client">  
          <prefs xmlns="urn:xmpp:mam:0"></prefs>  
     </iq>  
</body>`

* and this is what i got in response from openfire:
`<body xmlns="http://jabber.org/protocol/httpbind" ack="550948740">  
     <iq id="1:sendIQ" xmlns="jabber:client" type="error" to="796@192.168.5.43/null">  
          <error code="501" type="cancel">  
               <feature-not-implemented xmlns="urn:ietf:params:xml:ns:xmpp-stanzas"></feature-not-implemented>  
          </error>  
     </iq>  
</body>`

I created discussion of this but its not getting approved: https://community.igniterealtime.org/thread/57111

**Pull request merged**